Add k-(node)-components algorithm.

jtorrents opened the pull request:
Add Moody and White algorithm for for identifying k-components in a graph,
which is based on Kanevsky's algorithm for finding all minimum-size node
cut-sets (implemented in `all_node_cuts` function #1391). This algorithm
consists in keep cutting a graph until we obtain either a complete or a
trivial graph.

`all_node_cuts` is doing most of the work here, the only tricky
implementation thing is the generation of partitions. The authors say that
``Nodes in a cutset belong to both sides of the induced cut.'', but an
induced cut can split a graph in more than two sides. Thus, after a cut,
we have to merge the node sets that (after adding the nodes in the cut set)
share at least k nodes.

I found out that this is very related to set consolidation. I've adapted a
nice function from  http://rosettacode.org/wiki/Set_consolidation . However
it's not clear the license of this implementation. I'll try to figure out.

jtorrents commented:
For reporting the results I'm using a dictionary with all connectivity levels `k` in the input Graph as keys and a list of sets of nodes that form a k-component of level `k` as values. This makes it very easy to build subgraphs that are k-components at any connectivity level. But requires a bit of work at the end to build a dictionary with all connectivity levels because k-components are only reported at their maximum k level.

jtorrents commented:
By the way, I'm using this nice graph with 99 nodes and 200 edges with a tricky connectivity structure for tests. It's a gird graph (with node connectivity 2), which has attached 4 Petersen graphs (with node connectivity 3) in the four corners. The four Petersen graphs, in turn, have attached two $k_5$ (with node connectivity 4), which share one or two nodes (see p.26 [this paper](http://arxiv.org/pdf/1503.04476v1) for more details.

![illustrative_knumber_exact](https://cloud.githubusercontent.com/assets/1184374/6698163/a4416060-ccf6-11e4-9c12-e243511b06b7.png)

jtorrents commented:
For completeness, here is the function that generates the graph above 

```python
def torrents_and_ferraro_graph():
    # Graph from http://arxiv.org/pdf/1503.04476v1 p.26
    G = nx.convert_node_labels_to_integers(nx.grid_graph([5,5]),
                                            label_attribute='labels')
    rlabels = nx.get_node_attributes(G, 'labels')
    labels = dict((v, k) for k, v in rlabels.items())

    for nodes in [ (labels[(0,4)], labels[(1,4)]),
                    (labels[(3,4)], labels[(4,4)]) ]:
        new_node = G.order()+1
        # Petersen graph is triconnected
        P = nx.petersen_graph()
        G = nx.disjoint_union(G,P)
        # Add two edges between the grid and P
        G.add_edge(new_node+1, nodes[0])
        G.add_edge(new_node, nodes[1])
        # K5 is 4-connected
        K = nx.complete_graph(5)
        G = nx.disjoint_union(G,K)
        # Add three edges between P and K5
        G.add_edge(new_node+2,new_node+11)
        G.add_edge(new_node+3,new_node+12)
        G.add_edge(new_node+4,new_node+13)
        # Add another K5 sharing a node
        G = nx.disjoint_union(G,K)
        nbrs = G[new_node+10]
        G.remove_node(new_node+10)
        for nbr in nbrs:
            G.add_edge(new_node+17, nbr)
        # Commenting this makes the graph not biconnected !!
        # This stupid mistake make one reviewer very angry :P
        G.add_edge(new_node+16, new_node+8)

    for nodes in [(labels[(0,0)], labels[(1,0)]),
                    (labels[(3,0)], labels[(4,0)])]:
        new_node = G.order()+1
        # Petersen graph is triconnected
        P = nx.petersen_graph()
        G = nx.disjoint_union(G,P)
        # Add two edges between the grid and P
        G.add_edge(new_node+1, nodes[0])
        G.add_edge(new_node, nodes[1])
        # K5 is 4-connected
        K = nx.complete_graph(5)
        G = nx.disjoint_union(G,K)
        # Add three edges between P and K5
        G.add_edge(new_node+2,new_node+11)
        G.add_edge(new_node+3,new_node+12)
        G.add_edge(new_node+4,new_node+13)
        # Add another K5 sharing two nodes
        G = nx.disjoint_union(G,K)
        nbrs = G[new_node+10]
        G.remove_node(new_node+10)
        for nbr in nbrs:
            G.add_edge(new_node+17, nbr)
        nbrs2 = G[new_node+9]
        G.remove_node(new_node+9)
        for nbr in nbrs2:
            G.add_edge(new_node+18, nbr)

    G.name = 'Example graph for connectivity'
    return G
```

And here is the result of applying this algorithm:

```python
In [6]: G = torrents_and_ferraro_graph()

In [7]: G_k = nx.k_components(G)

In [8]: G_k[3]
Out[8]: 
[{25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 42},
 {44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 61},
 {63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 79, 80},
 {81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 93, 94, 95, 99, 100},
 {39, 40, 41, 42, 43},
 {58, 59, 60, 61, 62},
 {76, 77, 78, 79, 80},
 {96, 97, 98, 99, 100}]

In [9]: G_k[4]
Out[9]: 
[{35, 36, 37, 38, 42},
 {39, 40, 41, 42, 43},
 {54, 55, 56, 57, 61},
 {58, 59, 60, 61, 62},
 {73, 74, 75, 79, 80},
 {76, 77, 78, 79, 80},
 {93, 94, 95, 99, 100},
 {96, 97, 98, 99, 100}]
```

Note that it's quite fast (here I only compare it with an older version of `igraph` via its python bindings, which is what I have installed in this machine right now, so I'm not claiming that this is a fair benchmark; it's only orientative):

```python
In [10]: %timeit G_k = nx.k_components(G)
1 loops, best of 3: 207 ms per loop
In [11]: import igraph
In [12]: print(igraph.__version__)
0.6.5
In [13]: from sna import utils
In [14]: g = utils.networkx_to_igraph(G)
In [15]: %timeit g.cohesive_blocks()
1 loops, best of 3: 11.5 s per loop
```

jtorrents commented:
Well I just tested this on the R bindings of igraph (still an older version), and it's quite faster than with the python bindings, but still slower than this implementation:

```R
> library(igraph)
> g <- read.graph('test.graphml', format='graphml')
> system.time(cb <- cohesive.blocks(g))
   user  system elapsed 
  1.628   0.000   1.629 
```

jtorrents commented:
It seems that Mathematica is also shipping a function for k-components in their last version:

http://reference.wolfram.com/language/ref/KVertexConnectedComponents.html

Does anybody has access to the last version of Mathematica? I'd love to check if it gets the example graph above right, and also have some speed benchmarks against our implementation.

jtorrents commented:
A few things on this. Regarding the code for the set consolidation from the Rosseta code project, it seems that the only license information is on the whole website and it's GNU/FDL (Free Documentation License). So it's not clear if that also covers the code included in the website. In any case here is the function:

```python
def _consolidate(sets, k):
    """Based on http://rosettacode.org/wiki/Set_consolidation """
    setlist = [s for s in sets if s]
    for i, s1 in enumerate(setlist):
        if s1:
            for s2 in setlist[i+1:]:
                intersection = s1.intersection(s2)
                if len(intersection) > k:
                    s2.update(s1)
                    s1.clear()
                    s1 = s2
    return [s for s in setlist if s]
```

I cannot think of another way to avoid two nested for loops (making it O(n^2)). Maybe there is another source for this implementation of set consolidation that have a license that we can use? If someone has ideas on this it would be much appreciated.

jtorrents commented:
I also did some speed test with the last released version of `igraph` (0.7.1). Igraph is much slower in bigger networks because their implementation of vertex connectivity does not use nice algorithms but just computes the local node connectivity on all pairs of nodes in the graph. See: 

https://github.com/igraph/igraph/blob/5cea4789827c9da2b424b65246d3b7d9642741e1/src/flow.c#L124

So it's not a very informative benchmark but here it goes anyway:

```python
from __future__ import print_function

from collections import OrderedDict
import time
import igraph
import networkx as nx
from networkx.algorithms.flow import edmonds_karp, shortest_augmenting_path
from networkx.utils import powerlaw_sequence, create_degree_sequence

def networkx_to_igraph(G, attrs=False):
    directed = G.is_directed()
    nodes = OrderedDict(G.nodes_iter(data=True))
    edges = OrderedDict([((u, v), d) for u, v, d in G.edges_iter(data=True)])
    graph = igraph.Graph(len(nodes), directed=directed)
    rids = dict(enumerate(nodes))
    ids = dict((v, k) for k, v in rids.items())
    for vertex in graph.vs:
        node = rids[vertex.index]
        if attrs:
            for k, v in nodes[node].items():
                vertex[k] = v
        vertex['id'] = node
    graph.add_edges([(ids[u], ids[v]) for u, v in edges])
    if attrs:
        for i, ((source, target), data) in enumerate(edges.items()):
            for k, v in data.items():
                graph.es[i][k] = v
    return graph


def build_power_law(n):
    deg_seq = create_degree_sequence(n, powerlaw_sequence, 100)
    G = nx.Graph(nx.configuration_model(deg_seq))
    G.remove_edges_from(G.selfloop_edges())
    G = max(nx.biconnected_component_subgraphs(G), key=len)
    G.name = 'Power law configuration model: {0}'.format(n)
    return G


def benchmark_connectivity():
    graphs = []
    for p in [0.1, 0.15, 0.2]:
        G = nx.fast_gnp_random_graph(100, p)
        graphs.append(G)
    for n in [500, 1000, 1500, 2000]:
        G = build_power_law(n)
        graphs.append(G)
    for G in graphs:
        libs = OrderedDict([
            ('NetworkX', G), 
            ('iGraph', networkx_to_igraph(G)),
        ])
        print(nx.info(G))
        for lib, graph in libs.items():
            print('Using {}:'.format(lib))
            if lib == 'NetworkX':
                flow_func = (shortest_augmenting_path if 'gnp' in graph.name
                                else edmonds_karp)
                print("\tComputing node connectivity")
                start = time.time()
                k = nx.node_connectivity(graph, flow_func=flow_func)
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
                print("\tComputing edge connectivity")
                start = time.time()
                k = nx.edge_connectivity(graph, flow_func=flow_func)
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
                print("\tComputing k connected components")
                start = time.time()
                k = nx.k_components(graph, flow_func=flow_func)
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
            if lib == 'iGraph':
                print("\tComputing node connectivity")
                start = time.time()
                k = graph.vertex_connectivity()
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
                print("\tComputing edge connectivity")
                start = time.time()
                k = graph.edge_connectivity()
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
                print("\tComputing k connected components")
                start = time.time()
                k = graph.cohesive_blocks()
                end = time.time() - start
                print("\tIt took {:.3f} seconds".format(end))
        print()


if __name__ == '__main__':
    benchmark_connectivity()
```

And the results:

```
jtorrents@saturn:~/projects/benchmark$ python3 benchmark_igraph_connectivity.py 
Name: fast_gnp_random_graph(100,0.1)
Type: Graph
Number of nodes: 100
Number of edges: 459
Average degree:   9.1800
Using NetworkX:
	Computing node connectivity
	It took 0.094 seconds
	Computing edge connectivity
	It took 0.012 seconds
	Computing k connected components
	It took 3.340 seconds
Using iGraph:
	Computing node connectivity
	It took 1.812 seconds
	Computing edge connectivity
	It took 0.005 seconds
	Computing k connected components
	It took 35.693 seconds

Name: fast_gnp_random_graph(100,0.15)
Type: Graph
Number of nodes: 100
Number of edges: 724
Average degree:  14.4800
Using NetworkX:
	Computing node connectivity
	It took 0.106 seconds
	Computing edge connectivity
	It took 0.015 seconds
	Computing k connected components
	It took 11.037 seconds
Using iGraph:
	Computing node connectivity
	It took 2.554 seconds
	Computing edge connectivity
	It took 0.007 seconds
	Computing k connected components
	It took 70.060 seconds

Name: fast_gnp_random_graph(100,0.2)
Type: Graph
Number of nodes: 100
Number of edges: 966
Average degree:  19.3200
Using NetworkX:
	Computing node connectivity
	It took 0.203 seconds
	Computing edge connectivity
	It took 0.019 seconds
	Computing k connected components
	It took 153.991 seconds
Using iGraph:
	Computing node connectivity
	It took 3.046 seconds
	Computing edge connectivity
	It took 0.009 seconds
	Computing k connected components
	It took 81.979 seconds

Name: Power law configuration model: 500
Type: Graph
Number of nodes: 328
Number of edges: 1095
Average degree:   6.6768
Using NetworkX:
	Computing node connectivity
	It took 0.223 seconds
	Computing edge connectivity
	It took 0.057 seconds
	Computing k connected components
	It took 6.185 seconds
Using iGraph:
	Computing node connectivity
	It took 59.922 seconds
	Computing edge connectivity
	It took 0.045 seconds
	Computing k connected components
	It took 150.375 seconds

Name: Power law configuration model: 1000
Type: Graph
Number of nodes: 638
Number of edges: 2376
Average degree:   7.4483
Using NetworkX:
	Computing node connectivity
	It took 0.930 seconds
	Computing edge connectivity
	It took 0.215 seconds
	Computing k connected components
	It took 31.760 seconds
Using iGraph:
	Computing node connectivity
	It took 480.676 seconds
	Computing edge connectivity
	It took 0.181 seconds
	Computing k connected components
	It took 1151.528 seconds

Name: Power law configuration model: 1500
Type: Graph
Number of nodes: 931
Number of edges: 3532
Average degree:   7.5875
Using NetworkX:
	Computing node connectivity
	It took 1.990 seconds
	Computing edge connectivity
	It took 0.465 seconds
	Computing k connected components
	It took 124.550 seconds
Using iGraph:
	Computing node connectivity
	It took 1556.392 seconds
	Computing edge connectivity
	It took 0.386 seconds
	Computing k connected components
^C
```
I just killed the benchmark at this point.

bjedwards commented:
@jtorrents I have access to Mathematica 10. I don't know a ton of Mathematica, but I tried your example out, and Mathematica doesn't look like it works for `k>2`. First the python

```python
import networkx as nx
def torrents_and_ferraro_graph():
#Copy and pasted
...

G = torrents_and_ferraro_graph()
nx.write_graph6(G,'t_and_f.graph6')
```
What I tried and got in Mathematica

```Mathematica
g = Import['t_and_f.graph6']
KVertexConnectedComponents[g,3]
KVertexConnectedComponents::ngen: The generalized KVertexConnectedComponents[Graph[<99>, <200>], 3] is not implemented. >>             1       CalculateScan`UnitScanner`Private`boxes A       •ê    p¢ï    •ê            omponents/Iconiz@   
```

For `k=2` it returns some output.

```Mathematica
In[11]:= KVertexConnectedComponents[g,2]
Out[11]= {{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99}}   
```

hagberg commented:
@jtorrents - with the set consolidation part could you use connected components or union find?

jtorrents commented:
@bjedwards, thanks for looking at this! Yes, judging by the error message, it seems that it's not implemented. In their [documentation](http://reference.wolfram.com/language/ref/KVertexConnectedComponents.html) they use the karate club graph as example, but only show results for k=2.

@hagberg Thanks! Indeed this problem can be solved using connected components. This is the implementation now:

```python
def _consolidate(sets, k):
    G = nx.Graph()
    nodes = {i: s for i, s in enumerate(sets)}
    G.add_nodes_from(nodes)
    G.add_edges_from((u, v) for u, v in combinations(nodes, 2)
                     if len(nodes[u] & nodes[v]) >= k)
    for component in nx.connected_components(G):
        yield set.union(*[nodes[n] for n in component])
```

This is slower than the iterative version from rosettacode.org because of the overhead of building a graph and calling nx.connected_components, but I think we perform the same number of
intersection operations.

Also, I found out that it's possible that subsets of an already detected k-component appear at a level k. Checking and acting on this in the main while loop penalizes the common case. Thus we also have to consolidate all connectivity levels in helper function _reconstruct_k_components. Improved tests to make sure that no spureous k-components are returned and added tests to `_consolidate` from examples in rosettacode.org.

jtorrents commented:
Improved tests, and added `k_components` to the API changes. Rebased to solve `api_2.0.rst` merge conflicts. I think this is ready for merging. Any comment welcome.

jtorrents commented:
Sorry for not acting on this before, I'm quite busy this days. I think this is ready and can be merged for 1.10. @ysitu do you think we should wait until 2.0 for this?

ysitu commented:
@jtorrents Your call. I did not follow this closely.

jtorrents commented:
I merged both this and its approximation counterpart.
